src/providers/on-chain-quote-provider.ts

```diff
@@ -944,4 +985,32 @@ export class V3QuoteProvider implements IV3QuoteProvider {
       );
     }
   }
+
+  /**
+   * Throw an error for incorrect routes / function combinations
+   * @param routes Any combination of V3, V2, and Mixed routes.
+   * @param functionName
+   * @param useQuoterV2 Boolean indicating whether the QuoterV2 needs to be used (for pure V3)
+   */
+  protected validateRoutes(
+    routes: (V3Route | V2Route | MixedRoute)[],
+    functionName: string,
+    useQuoterV2: boolean
+  ) {
+    // we cannot have both V3 and MixedRoutes or V2Routes in the same call since they use different quoter contracts
+    if (
+      !useQuoterV2 &&
+      routes.some((route) => route.protocol === Protocol.V3)
+    ) {
+      throw new Error(
+        'Cannot have mix and match V3 on chain quotes with MixedRoutes and/or V2Routes'
+      );
+    }
```

Comment:
I still find the comment and logic here super confusing.

If we are using mixedroute quoter, it seems fine to me that some of the routes are V3 only so not sure why we error out.

This would be cleaner to me:

```
if (useQuoterV2 && (routes.some(Protocol.V2) || routes.some(Protocol.MIXED)) {
    error
}
```